File: specter_exec/spawn.py
"""Run a subprocess to completion, streaming its output under a deadline.

The mechanics of running an agent, separated from knowing which agents exist.
Given a command, it streams stdout line by line, drains stderr on a thread, kills
the process if it overruns, and returns what happened.

WHY A THREAD FOR STDERR
A subprocess writing more to stderr than the pipe buffer holds blocks forever
while the parent reads stdout. Draining both concurrently is what stops a chatty
agent deadlocking a run.

WHY THE DEADLINE IS CHECKED WHILE READING
Between lines is the only place a streaming reader can notice time passing.
subprocess.run(timeout=) would be simpler but cannot stream, and streaming is the
point: a run that takes minutes has to show progress while it works.

Run:  python3 specter_exec/spawn.py --self-check
"""

from __future__ import annotations

import subprocess
import sys
import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable

# Enough to diagnose a failure without storing a whole build log in a DB row.
STDOUT_LIMIT = 20000
STDERR_LIMIT = 12000
# One progress line is a line, not a file.
LINE_LIMIT = 2000


@dataclass
class SpawnResult:
    """What happened. No transport shape — the caller decides how to report it."""

    exit_code: int | None
    stdout: str
    stderr: str
    timed_out: bool = False
    error: str = ""          # set only when the process could not start
    lines: list[str] = field(default_factory=list)

    @property
    def ok(self) -> bool:
        return self.exit_code == 0 and not self.timed_out and not self.error


def run_streaming(
    command: list[str],
    cwd: str | Path,
    timeout_seconds: int,
    on_stdout: Callable[[str], None] | None = None,
    on_stderr: Callable[[str], None] | None = None,
    on_proc: Callable[[subprocess.Popen], None] | None = None,
) -> SpawnResult:
    """Run `command` in `cwd`, streaming each line as it arrives.

    on_proc receives the live process before any output — that is what makes
    cancellation possible. Without it a caller can only wait for the deadline,
    which is the difference between a stoppable run and a stuck one.

    A failure to spawn returns a result with `error` set rather than raising:
    "claude is not installed" and "claude exited 1" are both outcomes the caller
    reports the same way.
    """
    stdout_lines: list[str] = []
    stderr_lines: list[str] = []
    timed_out = False

    try:
        proc = subprocess.Popen(
            command,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            bufsize=1,
            cwd=str(cwd),
        )
    except Exception as exc:  # noqa: BLE001 - missing binary, bad cwd, no permission
        return SpawnResult(exit_code=None, stdout="", stderr="", error=str(exc))

    if on_proc:
        on_proc(proc)

    def drain_stderr() -> None:
        for raw in proc.stderr:  # type: ignore[union-attr]
            line = raw.rstrip()
            stderr_lines.append(line)
            if on_stderr and line:
                on_stderr(line[:LINE_LIMIT])

    stderr_thread = threading.Thread(target=drain_stderr, daemon=True)
    stderr_thread.start()

    deadline = time.monotonic() + timeout_seconds
    try:
        for raw in proc.stdout:  # type: ignore[union-attr]
            line = raw.rstrip()
            stdout_lines.append(line)
            if on_stdout:
                on_stdout(line)
            if time.monotonic() > deadline:
                proc.kill()
                timed_out = True
                break
        proc.wait()
        # Bounded: a stderr thread blocked on a pipe must not hold the run open.
        stderr_thread.join(timeout=2)
    except Exception as exc:  # noqa: BLE001
        try:
            proc.kill()
        except Exception:  # noqa: BLE001
            pass
        return SpawnResult(exit_code=None, stdout="\n".join(stdout_lines),
                           stderr="\n".join(stderr_lines), error=str(exc))

    return SpawnResult(
        exit_code=proc.returncode,
        stdout="\n".join(stdout_lines)[-STDOUT_LIMIT:],
        stderr="\n".join(stderr_lines)[-STDERR_LIMIT:],
        timed_out=timed_out,
        lines=stdout_lines,
    )


# ── self-check ───────────────────────────────────────────────────────────────

def _demo() -> None:
    import tempfile

    failed = 0

    def check(name: str, got, want) -> None:
        nonlocal failed
        ok = got == want
        print(f"  {'✓' if ok else '✗'} {name}" + ("" if ok else f"  got {got!r}, want {want!r}"))
        if not ok:
            failed += 1

    with tempfile.TemporaryDirectory() as tmp:
        r = run_streaming(["echo", "hello"], tmp, 10)
        check("captures stdout", r.stdout, "hello")
        check("reports success", r.ok, True)
        check("exit code 0", r.exit_code, 0)

        r = run_streaming(["sh", "-c", "exit 3"], tmp, 10)
        check("captures a non-zero exit", r.exit_code, 3)
        check("and is not ok", r.ok, False)

        r = run_streaming(["sh", "-c", "echo oops >&2"], tmp, 10)
        check("captures stderr separately", r.stderr, "oops")

        seen: list[str] = []
        run_streaming(["sh", "-c", "echo one; echo two"], tmp, 10, on_stdout=seen.append)
        check("streams each line as it arrives", seen, ["one", "two"])

        # THE DEADLOCK CASE. More stderr than the pipe buffer holds: without a
        # draining thread this hangs forever rather than failing.
        # Bounded generator, not `yes | head`: `yes` never exits on its own, so
        # that version left a process spinning after head closed the pipe --
        # my own test hung, not the module.
        r = run_streaming(
            ["python3", "-c",
             "import sys\nfor i in range(20000): sys.stderr.write('noise line %d\\n' % i)"],
            tmp, 15)
        check("a chatty stderr does not deadlock", r.exit_code, 0)

        # THE DEADLINE. Must kill, not wait for the process to finish.
        start = time.monotonic()
        r = run_streaming(["sh", "-c", "for i in 1 2 3 4 5 6 7 8; do echo tick; sleep 1; done"],
                          tmp, 2)
        elapsed = time.monotonic() - start
        check("overrunning is reported as a timeout", r.timed_out, True)
        check("and is not ok", r.ok, False)
        check("and does not wait for the full command", elapsed < 6, True)

        # A missing binary is an outcome, not an exception to handle upstream.
        r = run_streaming(["definitely-not-a-real-binary-xyz"], tmp, 5)
        check("a missing binary returns an error result", bool(r.error), True)
        check("rather than raising", r.ok, False)
        check("with no exit code", r.exit_code, None)

        # CANCELLATION: on_proc must fire before output, or a caller has nothing
        # to kill while the run is in flight.
        captured = {}
        run_streaming(["echo", "x"], tmp, 10, on_proc=lambda p: captured.setdefault("proc", p))
        check("the live process is handed to the caller", "proc" in captured, True)

    print(f"\n  {'FAILED' if failed else 'all checks passed'}"
          + (f" — {failed} assertion(s)" if failed else ""))
    raise SystemExit(1 if failed else 0)


if __name__ == "__main__" and "--self-check" in sys.argv:
    _demo()

File: internal/exec/spawn.go
package exec

import (
	"bufio"
	"context"
	"errors"
	"os/exec"
	"strings"
	"sync"
	"time"
)

// Running a subprocess to completion, streaming its output under a deadline.
//
// The mechanics of running an agent, separated from knowing which agents exist.
//
// WHY BOTH PIPES ARE DRAINED CONCURRENTLY
// A subprocess writing more to stderr than the pipe buffer holds blocks forever
// while the parent reads stdout. Draining both at once is what stops a chatty
// agent deadlocking a run.
//
// WHY exec.CommandContext RATHER THAN A MANUAL DEADLINE
// The Python version checked the clock between stdout lines, which only notices
// time passing while output is arriving — a silent agent could overrun. A context
// deadline fires regardless, and the same mechanism serves cancellation.

// drainGrace is how long output collection may continue after the process has
// exited. Long enough for a normal flush, short enough that a pipe held by a
// surviving grandchild cannot hold the run open.
const drainGrace = 2 * time.Second

const (
	// Enough to diagnose a failure without storing a build log in a database row.
	StdoutLimit = 20000
	StderrLimit = 12000
)

// Command is what to run.
type Command struct {
	Argv     []string
	Dir      string
	Env      []string
	Timeout  time.Duration
	OnStdout func(string)
	OnStderr func(string)
}

// Result is what happened. No transport shape — the caller decides how to report.
type Result struct {
	ExitCode int      `json:"exit_code"`
	Stdout   string   `json:"stdout"`
	Stderr   string   `json:"stderr"`
	TimedOut bool     `json:"timed_out"`
	Err      string   `json:"error,omitempty"` // set only when the process could not run
	Lines    []string `json:"-"`
}

func (r Result) OK() bool {
	return r.ExitCode == 0 && !r.TimedOut && r.Err == ""
}

// RunStreaming runs cmd, streaming each line as it arrives.
//
// The context governs cancellation; Timeout adds a deadline on top. Either
// stopping the run is reported as TimedOut, because from the caller's side they
// are the same outcome: it did not finish on its own.
//
// A failure to start returns a Result with Err set rather than an error return:
// "claude is not installed" and "claude exited 1" are both outcomes a caller
// reports the same way.
func RunStreaming(ctx context.Context, cmd Command) Result {
	if cmd.Timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, cmd.Timeout)
		defer cancel()
	}

	proc := exec.CommandContext(ctx, cmd.Argv[0], cmd.Argv[1:]...)
	proc.Dir = cmd.Dir
	if cmd.Env != nil {
		proc.Env = cmd.Env
	}

	stdout, err := proc.StdoutPipe()
	if err != nil {
		return Result{ExitCode: -1, Err: err.Error()}
	}
	stderr, err := proc.StderrPipe()
	if err != nil {
		return Result{ExitCode: -1, Err: err.Error()}
	}

	if err := proc.Start(); err != nil {
		// Missing binary, bad working directory, no permission.
		return Result{ExitCode: -1, Err: err.Error()}
	}

	var (
		mu          sync.Mutex
		stdoutLines []string
		stderrLines []string
		wg          sync.WaitGroup
	)

	drain := func(r *bufio.Scanner, sink *[]string, on func(string)) {
		defer wg.Done()
		// Agent output can exceed bufio's default 64 KiB line cap; a long line
		// would otherwise end the scan early and truncate the run's output.
		r.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
		for r.Scan() {
			line := strings.TrimRight(r.Text(), "\r\n")
			mu.Lock()
			*sink = append(*sink, line)
			mu.Unlock()
			if on != nil {
				on(line)
			}
		}
	}

	wg.Add(2)
	go drain(bufio.NewScanner(stdout), &stdoutLines, cmd.OnStdout)
	go drain(bufio.NewScanner(stderr), &stderrLines, cmd.OnStderr)

	// Wait for the drains, but NOT unconditionally.
	//
	// Killing the child does not close a pipe its own child inherited. An agent
	// that leaves a background process behind keeps the write end open, so
	// scanning never reaches EOF and this would block past the deadline --
	// forever, with nothing reporting why. Waiting on the PROCESS and giving the
	// drains a short grace period afterwards bounds it: output already written
	// is still collected, and a pipe nobody will close no longer holds the run.
	drained := make(chan struct{})
	go func() { wg.Wait(); close(drained) }()

	// Drains FIRST, then Wait. os/exec closes both pipes inside Wait, and its
	// own documentation says it is incorrect to call Wait before the reads from
	// those pipes have finished. Calling it first is a race that only loses
	// under load: a script that writes and exits immediately can have its pipes
	// closed before the drain goroutines are ever scheduled, and the output is
	// gone. On an idle machine the drains win and nothing looks wrong.
	//
	// That produced failures nowhere near the cause — an agent's reply of "YES"
	// classified as "no", a memory row never written, a second node running
	// with no context — all of them "the output was empty", in three different
	// packages, only on a busy CI runner.
	//
	// The deadline stays, for the reason it was added: killing the child does
	// not close a pipe its own child inherited, so an agent that leaves a
	// background process behind holds the write end open and the scan never
	// reaches EOF. Bounding the wait means output already written is still
	// collected, while a pipe nobody will close cannot hold the run forever.
	select {
	case <-drained:
	case <-time.After(drainGrace):
	}

	waitErr := proc.Wait()

	mu.Lock()
	outText := strings.Join(stdoutLines, "\n")
	errText := strings.Join(stderrLines, "\n")
	lines := append([]string(nil), stdoutLines...)
	mu.Unlock()

	result := Result{
		Stdout: tail(outText, StdoutLimit),
		Stderr: tail(errText, StderrLimit),
		Lines:  lines,
	}

	// Reading the collected lines while a drain goroutine may still be appending
	// to them: the mutex makes that safe, and an orphaned drain writing into a
	// slice nobody reads again is harmless.

	// Context death means the deadline fired or the caller cancelled. Either way
	// the run did not finish on its own, and the exit code reflects the signal
	// rather than anything the agent decided.
	if ctxErr := ctx.Err(); ctxErr != nil {
		result.TimedOut = true
		result.ExitCode = -1
		return result
	}

	if waitErr != nil {
		var exitErr *exec.ExitError
		if errors.As(waitErr, &exitErr) {
			result.ExitCode = exitErr.ExitCode()
		} else {
			result.ExitCode = -1
			result.Err = waitErr.Error()
		}
		return result
	}

	result.ExitCode = 0
	return result
}

// tail keeps the last n bytes — the end of a failing run is what diagnoses it.
func tail(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[len(s)-n:]
}
